csrc/preseg_passes/exact_mapped_extent_substitution.cpp

```diff
@@ -22,8 +22,51 @@ namespace {
 // otherwise, validateDomainEquivalence fails. If we really want to substitute,
 // we may need to skip or modify validateDomainEquivalence.
 inline bool isNonSubstitutableID(const IterDomain* id) {
-  return (id->isBroadcast() && !id->hasExpandedExtent()) || id->definition() ||
-      id->extent()->definition();
+  return id == nullptr || (id->isBroadcast() && !id->hasExpandedExtent()) ||
+      id->definition() || id->extent()->definition();
+}
+
+// Build disjoint set of extents from  disjoint set of ids
+auto buildExtentSetFromIdSets(const DisjointSets<Val*>& id_sets) {
+  DisjointSets<Val*> extent_sets;
+  // Loop over each id set
+  for (const auto& set_ptr : id_sets.disjointSets()) {
+    // If one of the extent in this set is already in the extent_sets, then
+    // map all other extents to the same set, otherwise create a new set.
+    DisjointSets<Val*>::DisjointSet current_set = nullptr;
+
+    // First loop over the set, to check if one of the extent is already mapped
+    for (auto v : *set_ptr) {
+      auto id = dynamic_cast<IterDomain*>(v);
+      if (isNonSubstitutableID(id)) {
+        continue;
+      }
+      if (extent_sets.mappingExists(id->extent())) {
+        current_set = extent_sets.disjointSetMap().at(id->extent());
+      }
+    }
+
+    // Second loop over the set, to map all extents to the same set.
+    for (auto v : *set_ptr) {
+      auto id = dynamic_cast<IterDomain*>(v);
+      if (isNonSubstitutableID(id)) {
+        continue;
+      }
+      // Here extent is used instead of expanded exent since a bcast dim may
+      // be expanded to different extents, e.g. issue-3227.
+      auto extent = id->extent();
+      if (extent_sets.mappingExists(extent)) {
+        continue;
+      }
+      if (current_set) {
+        extent_sets.appendToSet(extent, current_set);
+      } else {
+        auto it = extent_sets.initializeSet(extent).first;
+        current_set = it->second;
```

Comment:
If we move these two lines after the first loop, i.e., do this if not existing set is found, wouldn't it make the second loop simpler?

By doing so, at the end of the first loop, it's guaranteed to have non-null `current_set`, so the second loop should just need to loop over the IDs and do `appedToSet` with no conditional check.

Reply:
Good point. Changed to `if current_extent_set is not set after the 1st loop, create a new set using the extent of the 1st id `
```
    // Create a new set if no extent is mapped
    if (current_extent_set == nullptr) {
      auto id = dynamic_cast<IterDomain*>(*id_set_ptr->begin());
      auto extent = id->extent();
      auto it = extent_sets.initializeSet(extent).first;
      current_extent_set = it->second;
    }

```